contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+statefulset controller will update this on the federated headless service while creating it, or if
+its applied/updated later, will update it accordingly.
+Additionally the information about the stateful replicas (and the same name LB services) for which the
+service controller should update the dns records against the headless service will also need to be passed
+on to the headless service. The proposal is to pass this information through annotations on the headless
+service as below:
+```
+federation.kubernetes.io/stateful-replica-list: <replica-name-1>, <replica-name-2>, ..., <replica-name-n>
+```
+The federation service controller, if finds the names in this list against a federated headless service,
+will look for matching federated LB services and update the CNAME record for the headless service with the
+relevant SRV records for all the LB services in the federation dns server.
+This list can certainly be updated to the first class API data, if the federation API objects and clients
+are updated and maintained separately to have additional fields/data then that present in k8s API.
+
+Once this happens, the stateful pods can discover peers by querying the governing service domain (federated
+headless service). Taking the same example listed earlier if the peer (or a user) queries the service
+```store.mynamespace.myfederation.svc.example.com```, the below SRV records will be returned.
+```
+store-c1-0.etcd.mynamespace.myfederation.svc.example.com
+store-c1-1.etcd.mynamespace.myfederation.svc.example.com
+store-c1-2.etcd.mynamespace.myfederation.svc.example.com
+store-c1-3.etcd.mynamespace.myfederation.svc.example.com
+store-c2-0.etcd.mynamespace.myfederation.svc.example.com
+store-c2-1.etcd.mynamespace.myfederation.svc.example.com
+store-c2-2.etcd.mynamespace.myfederation.svc.example.com
+store-c2-3.etcd.mynamespace.myfederation.svc.example.com
+store-c3-0.etcd.mynamespace.myfederation.svc.example.com
+store-c3-1.etcd.mynamespace.myfederation.svc.example.com
+store-c3-2.etcd.mynamespace.myfederation.svc.example.com
+```
+The global federation dns service will have SRV records and relevant A records listing IPs of the LBs.
+The local in-cluster kube-dns will have the SRV records and the relevant A records listing the local
+stateful set pod IPs. The local kube-dns will be able to respond with both the local and global records,
+when a query hits the local kube-dns. The peers can choose to connect to the local identities or the
+global ones as the need be.
+
+#### Handling additional LB services
+
+As listed in the above approach, the current option of exposing the statefulset pods across clusters is to
+assign a LB service to each replica pod.
+If the federated service controller is used (by creating the service in federation) to create these services,
+then as per the current behaviour the service will be created in each cluster.
+This leads to exactly one unused service per statefulset per cluster.
```

Comment:
Actually that is not what I did intend to state. The default behaviour of federated service creation is create the same object in all federated clusters. 
So for example if the service named etcd-cluster1-0 is created in federation, all clusters will have one service each with this name.
The cluster selector now facilitates the behaviour needed for this proposal. 
Did you mean to say, that the service will be of use only in the cluster hosting the pod ?